File: plugins/plugin-sql/rust/src/schema/memory.rs
#![allow(missing_docs)]
//! Memory schema for elizaOS database
//!
//! Corresponds to the TypeScript memoryTable in packages/plugin-sql/typescript/schema/memory.ts

/// SQL for creating the memories table
pub const CREATE_MEMORIES_TABLE: &str = r#"
CREATE TABLE IF NOT EXISTS memories (
    id UUID PRIMARY KEY NOT NULL,
    type TEXT NOT NULL,
    created_at TIMESTAMPTZ NOT NULL DEFAULT now(),
    content JSONB NOT NULL,
    entity_id UUID REFERENCES entities(id) ON DELETE CASCADE,
    agent_id UUID REFERENCES agents(id) ON DELETE CASCADE NOT NULL,
    room_id UUID REFERENCES rooms(id) ON DELETE CASCADE,
    world_id UUID,
    "unique" BOOLEAN NOT NULL DEFAULT true,
    metadata JSONB NOT NULL DEFAULT '{}'
)
"#;

/// SQL for creating indexes on memories table
pub const CREATE_MEMORIES_INDEXES: &str = r#"
CREATE INDEX IF NOT EXISTS idx_memories_type_room ON memories (type, room_id);
CREATE INDEX IF NOT EXISTS idx_memories_world_id ON memories (world_id);
CREATE INDEX IF NOT EXISTS idx_memories_metadata_type ON memories ((metadata->>'type'));
CREATE INDEX IF NOT EXISTS idx_memories_document_id ON memories ((metadata->>'documentId'));
CREATE INDEX IF NOT EXISTS idx_fragments_order ON memories ((metadata->>'documentId'), (metadata->>'position'));
"#;

/// SQL for memory constraints
pub const CREATE_MEMORIES_CONSTRAINTS: &str = r#"
ALTER TABLE memories 
ADD CONSTRAINT IF NOT EXISTS fragment_metadata_check CHECK (
    CASE 
        WHEN metadata->>'type' = 'fragment' THEN
            metadata ? 'documentId' AND 
            metadata ? 'position'
        ELSE true
    END
);

ALTER TABLE memories 
ADD CONSTRAINT IF NOT EXISTS document_metadata_check CHECK (
    CASE 
        WHEN metadata->>'type' = 'document' THEN
            metadata ? 'timestamp'
        ELSE true
    END
);
"#;

/// Memory record structure for database operations
#[derive(Clone, Debug)]
pub struct MemoryRecord {
    pub id: uuid::Uuid,
    pub memory_type: String,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub content: serde_json::Value,
    pub entity_id: Option<uuid::Uuid>,
    pub agent_id: uuid::Uuid,
    pub room_id: Option<uuid::Uuid>,
    pub world_id: Option<uuid::Uuid>,
    pub unique: bool,
    pub metadata: serde_json::Value,
}

impl MemoryRecord {
    /// Convert to elizaOS Memory type
    pub fn to_memory(&self) -> elizaos::Memory {
        use elizaos::{Content, Memory, MemoryMetadata, UUID};

        let content: Content = serde_json::from_value(self.content.clone()).unwrap_or_default();

        let metadata: Option<MemoryMetadata> =
            if self.metadata.is_object() && !self.metadata.as_object().unwrap().is_empty() {
                serde_json::from_value(self.metadata.clone()).ok()
            } else {
                None
            };

        Memory {
            id: Some(UUID::new(&self.id.to_string()).unwrap()),
            entity_id: UUID::new(&self.entity_id.map(|u| u.to_string()).unwrap_or_default())
                .unwrap_or_else(|_| UUID::new_v4()),
            agent_id: Some(UUID::new(&self.agent_id.to_string()).unwrap()),
            created_at: Some(self.created_at.timestamp_millis()),
            content,
            embedding: None, // Loaded from embeddings table
            room_id: UUID::new(&self.room_id.map(|u| u.to_string()).unwrap_or_default())
                .unwrap_or_else(|_| UUID::new_v4()),
            world_id: self.world_id.map(|u| UUID::new(&u.to_string()).unwrap()),
            unique: Some(self.unique),
            similarity: None,
            metadata,
        }
    }

    /// Convert from elizaOS Memory type
    pub fn from_memory(memory: &elizaos::Memory, table_name: &str) -> Self {
        MemoryRecord {
            id: memory
                .id
                .as_ref()
                .map(|u| uuid::Uuid::parse_str(u.as_str()).unwrap())
                .unwrap_or_else(uuid::Uuid::new_v4),
            memory_type: table_name.to_string(),
            created_at: memory
                .created_at
                .map(|ts| chrono::DateTime::from_timestamp_millis(ts).unwrap())
                .unwrap_or_else(chrono::Utc::now),
            content: serde_json::to_value(&memory.content).unwrap_or_default(),
            entity_id: Some(uuid::Uuid::parse_str(memory.entity_id.as_str()).unwrap()),
            agent_id: memory
                .agent_id
                .as_ref()
                .map(|u| uuid::Uuid::parse_str(u.as_str()).unwrap())
                .unwrap_or_else(uuid::Uuid::new_v4),
            room_id: Some(uuid::Uuid::parse_str(memory.room_id.as_str()).unwrap()),
            world_id: memory
                .world_id
                .as_ref()
                .map(|u| uuid::Uuid::parse_str(u.as_str()).unwrap()),
            unique: memory.unique.unwrap_or(true),
            metadata: serde_json::to_value(&memory.metadata).unwrap_or_default(),
        }
    }
}

File: plugins/plugin-sql/src/schema/memory.ts
/**
 * Core memory table: every message, fact, document, or fragment an agent
 * stores, keyed by `type` and free-form `content`/`metadata` JSON. Cascade
 * deletes with its room, entity, or agent. Partial expression indexes and
 * CHECK constraints on `metadata->>'type'` enforce shape invariants for the
 * `fragment` and `document` metadata kinds without a dedicated column per
 * kind. Relations are defined in `embedding.ts` to avoid a circular import.
 */
import { type SQL, type SQLWrapper, sql } from "drizzle-orm";
import {
  boolean,
  check,
  foreignKey,
  index,
  jsonb,
  pgTable,
  text,
  timestamp,
  uuid,
} from "drizzle-orm/pg-core";
import { agentTable } from "./agent";
import { entityTable } from "./entity";
import { roomTable } from "./room";

/**
 * Produces exact search tokens or bounded GIN candidates. Candidate pieces can
 * match part of a longer word, so callers must also apply the exact predicate.
 * Only the index representation is split; stored and returned text stays whole.
 */
export function documentSearchQueryTokensExpression(
  input: SQLWrapper,
  indexCandidates = false
): SQL {
  const normalized = sql`translate(
    trim(${input}),
    'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
    'abcdefghijklmnopqrstuvwxyz'
  )`;
  // PostgreSQL GIN entries must fit an index page. 128 Unicode characters
  // occupy at most 512 UTF-8 bytes; the complete token remains in the source.
  const indexed = indexCandidates
    ? sql`regexp_replace(${normalized}, E'([^ \\t\\r\\n\\f]{128})', E'\\1 ', 'g')`
    : normalized;
  return sql`regexp_split_to_array(trim(${indexed}), E'[ \\t\\r\\n\\f]+')`;
}

export function documentSearchTokensExpression(
  content: SQLWrapper,
  metadata: SQLWrapper,
  indexCandidates = false
): SQL {
  return documentSearchQueryTokensExpression(
    sql`
    COALESCE(${content}->>'text', '') || E'\n' ||
    COALESCE(${metadata}->>'title', '') || E'\n' ||
    COALESCE(${metadata}->>'filename', '') || E'\n' ||
    COALESCE(${metadata}->>'originalFilename', '') || E'\n' ||
    COALESCE(${metadata}->>'source', '')
  `,
    indexCandidates
  );
}

export const memoryTable = pgTable(
  "memories",
  {
    id: uuid("id").primaryKey().notNull(),
    type: text("type").notNull(),
    createdAt: timestamp("created_at").default(sql`now()`).notNull(),
    content: jsonb("content").notNull(),
    entityId: uuid("entity_id").references(() => entityTable.id, {
      onDelete: "cascade",
    }),
    agentId: uuid("agent_id")
      .references(() => agentTable.id, {
        onDelete: "cascade",
      })
      .notNull(),
    roomId: uuid("room_id").references(() => roomTable.id, {
      onDelete: "cascade",
    }),
    worldId: uuid("world_id"),
    unique: boolean("unique").default(true).notNull(),
    metadata: jsonb("metadata").default({}).notNull(),
  },
  (table) => [
    index("idx_memories_type_room").on(table.type, table.roomId),
    index("idx_memories_world_id").on(table.worldId),
    foreignKey({
      name: "fk_room",
      columns: [table.roomId],
      foreignColumns: [roomTable.id],
    }).onDelete("cascade"),
    foreignKey({
      name: "fk_user",
      columns: [table.entityId],
      foreignColumns: [entityTable.id],
    }).onDelete("cascade"),
    foreignKey({
      name: "fk_agent",
      columns: [table.agentId],
      foreignColumns: [agentTable.id],
    }).onDelete("cascade"),
    index("idx_memories_metadata_type").on(sql`((metadata->>'type'))`),
    index("idx_memories_document_id").on(sql`((metadata->>'documentId'))`),
    index("idx_memories_document_search")
      .using("gin", documentSearchTokensExpression(table.content, table.metadata, true))
      .where(sql`${table.type} = 'documents' AND ${table.metadata}->>'type' = 'document'`),
    index("idx_fragments_order").on(
      sql`((metadata->>'documentId'))`,
      sql`((metadata->>'position'))`
    ),
    check(
      "fragment_metadata_check",
      sql`
            CASE 
                WHEN metadata->>'type' = 'fragment' THEN
                    metadata ? 'documentId' AND 
                    metadata ? 'position'
                ELSE true
            END
        `
    ),
    check(
      "document_metadata_check",
      sql`
            CASE 
                WHEN metadata->>'type' = 'document' THEN
                    metadata ? 'timestamp'
                ELSE true
            END
        `
    ),
  ]
);
